File: backend/app/schemas/specification.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from .common import LlmDebugInfo
from .sections import DocumentSection


class SpecificationAnchor(BaseModel):
    index: int = Field(..., description="Позиция блока в документе, начиная с 0")
    type: Literal["paragraph", "table"] = Field(..., description="Тип блока")
    preview: str = Field(..., description="Короткий фрагмент текста для ориентирования")


class SpecificationTable(BaseModel):
    index: int = Field(..., description="Позиция таблицы в документе")
    row_count: int = Field(..., description="Количество строк таблицы")
    column_count: int = Field(..., description="Количество столбцов таблицы")
    preview: str = Field(..., description="Первый ряд таблицы или ключевые ячейки")
    start_anchor: SpecificationAnchor = Field(..., description="Якорь начала таблицы")
    end_anchor: SpecificationAnchor = Field(..., description="Якорь конца таблицы")
    rows: list[list[str]] = Field(..., description="Содержимое таблицы построчно")


class SpecificationResponse(BaseModel):
    heading: str = Field(..., description="Заголовок раздела 'Спецификация'")
    start_anchor: SpecificationAnchor = Field(..., description="Начальная точка раздела")
    end_anchor: SpecificationAnchor = Field(..., description="Конечная точка раздела")
    tables: list[SpecificationTable] = Field(..., description="Обнаруженные таблицы раздела")


class SpecificationExtractionResponse(BaseModel):
    specification: SpecificationResponse = Field(..., description="Результаты извлечения")
    debug: LlmDebugInfo | None = Field(
        default=None,
        description="Диагностика, если документ обрабатывался через ИИ",
    )
    exported_json_name: str | None = Field(
        default=None,
        description="Имя JSON-файла с нормализованными позициями спецификации",
    )
    exported_json_base64: str | None = Field(
        default=None,
        description="Содержимое JSON-файла в кодировке base64",
    )
    sections: list[DocumentSection] = Field(
        default_factory=list,
        description="Найденные разделы контракта (для отображения в интерфейсе)",
    )
    combined_sections_name: str | None = Field(
        default=None,
        description="Имя объединенного txt-файла с шапкой и разделами",
    )
    combined_sections_base64: str | None = Field(
        default=None,
        description="Содержимое объединенного txt-файла в base64",
    )
    combined_sections_text: str | None = Field(
        default=None,
        description="Текстовое содержимое объединенного файла",
    )


class CroppedSpecResponse(BaseModel):
    specification: SpecificationResponse = Field(...)
    cropped_file_base64: str = Field(..., description="DOCX файла, base64 без префикса data:")
    cropped_file_name: str = Field(..., description="Имя файла для сохранения на клиенте")

File: frontend/src/api/specification.ts
import { resolveApiUrl } from "./client";

export type SpecificationMode = "ai" | "internal";

export interface LlmDebugInfo {
  prompt: Array<{ role: string; content: string }>;
  prompt_formatted: string;
  response: Record<string, unknown>;
  response_formatted: string;
}

export interface SpecificationAnchor {
  index: number;
  type: "paragraph" | "table";
  preview: string;
}

export interface SpecificationTable {
  index: number;
  row_count: number;
  column_count: number;
  preview: string;
  start_anchor: SpecificationAnchor;
  end_anchor: SpecificationAnchor;
  rows: string[][];
}

export interface DocumentSection {
  number: number | null;
  title: string;
  content: string;
  filename?: string | null;
}

export interface SpecificationResponse {
  heading: string;
  start_anchor: SpecificationAnchor;
  end_anchor: SpecificationAnchor;
  tables: SpecificationTable[];
}

export interface SpecificationExtractionResponse {
  specification: SpecificationResponse;
  debug?: LlmDebugInfo | null;
  exported_json_name?: string | null;
  exported_json_base64?: string | null;
  sections?: DocumentSection[];
  combined_sections_name?: string | null;
  combined_sections_base64?: string | null;
  combined_sections_text?: string | null;
}

export interface SpecificationFileResponse {
  specification: SpecificationResponse;
  cropped_file_name: string;
  cropped_file_base64: string;
}

export interface SectionReview {
  title: string;
  resume: string;
  risks: string;
  score: string;
}

export interface SectionReviewResponse {
  reviews: SectionReview[];
  overall_score?: number | null;
  red_flags?: string | null;
  html: string;
  debug?: LlmDebugInfo | null;
}

export interface FullProcessingResponse {
  specification_text?: string | null;
  docx_text?: string | null;
  overall_score?: number | null;
  inaccuracy?: string | null;
  red_flags?: string | null;
  html: string;
  debug?: LlmDebugInfo | null;
}

export async function uploadSpecificationDocument(
  file: File,
  mode: SpecificationMode,
  prompt?: string,
): Promise<SpecificationExtractionResponse> {
  const formData = new FormData();
  formData.append("file", file);
  if (prompt && prompt.trim()) {
    formData.append("prompt", prompt.trim());
  }

  const endpoint = mode === "ai" ? "/api/specification/ai" : "/api/specification/internal";

  const response = await fetch(resolveApiUrl(endpoint), {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const payload = await response.json().catch(() => null);
    const message = payload?.detail ?? payload?.error ?? "Не удалось обработать документ";
    throw new Error(message);
  }

  return (await response.json()) as SpecificationExtractionResponse;
}

export async function uploadInstructionFile(
  file: File,
): Promise<SectionReviewResponse> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await fetch(resolveApiUrl("/api/sections/review"), {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const payload = await response.json().catch(() => null);
    const message = payload?.detail ?? payload?.error ?? "Не удалось обработать файл";
    throw new Error(message);
  }

  return (await response.json()) as SectionReviewResponse;
}

export async function processFullContract(
  file: File,
): Promise<FullProcessingResponse> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await fetch(resolveApiUrl("/api/sections/full"), {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const payload = await response.json().catch(() => null);
    const message = payload?.detail ?? payload?.error ?? "Не удалось обработать договор";
    throw new Error(message);
  }

  return (await response.json()) as FullProcessingResponse;
}